### ISSUE TITLE
4.0rc1 qui-devices & qui-domains multi-start

### ISSUE THREAD

**Issue by 3hhh** [QubesOS/qubes-issues#3031]
The systray apps added in 4.0rc1 do not check at startup whether they are already started.

Thus e.g. a lightdm restart will cause the apps to be started again.

**Comment by 3hhh** [comment]
/etx/xdg/autostart/qui-devices.desktop
/etx/xdg/autostart/qui-domains.desktop

**Comment by marmarek** [comment]
The problem is not (lack of) check at startup, but rather not exiting
when you kill your session. I can't check it right now, but I'm almost
sure that they do exit when you terminate X server...

Anyway, this isn't the only thing that may be broken by restart of X
server in dom0. Other examples: 
 - qvm-start-gui have the same problem
 - audio from VM do not recover after it
 - probably more...

So, while there is some code to recover VM GUI after such event (so your
applications running in VM survive X restart), I'd still treat X server
restart as something that should be avoided if possible.

-- 
Best Regards,
Marek Marczykowski-Górecki
Invisible Things Lab
A: Because it messes up the order in which people normally read text.
Q: Why is top-posting such a bad thing?

**Comment by 3hhh** [comment]
I renamed it accordingly.

Just also noticed that dispVM instances are not deleted on `systemctl restart lightdm`.

**Comment by 3hhh** [comment]
I agree that it's not common user behavior and should thus only be low prio though.

**Comment by 3hhh** [comment]
@3hhh

> > Thus e.g. a lightdm restart will cause the apps to be started again.

> How do you know they're started? Did you see them in ps(1)?

Yes, exactly. `ps aux | grep python3` or so. Unfortunately I currently cannot test it again as they're not started anymore at all for some reason.

> > Just also noticed that dispVM instances are not deleted on systemctl restart lightdm.

> It's because they're X client instances are not killed and can be recovered by starting qvm-start --gui

Hm ok, that's expected. It would be nice if they'd be cleaned up during Qubes shutdown though as they tend to survive a laptop reboot if I recall correctly.
Unfortunately the last testing repo pull also broke my dispoable VMs so I cannot test that atm neither, sorry...

I'll provide an update once my system is more stable again...

**Comment by kalkin** [comment]
> It would be nice if they'd be cleaned up during Qubes shutdown though as they tend to survive a laptop reboot if I recall correctly

They can't survive a reboot. They only survive a X-server restart.

**Comment by 3hhh** [comment]
> They can't survive a reboot. They only survive a X-server restart.

I'm a tester, not a believer...

Anyway the dispVM issue does not seem to be related to X server restarts - thus I created #3037 for that one.

@**Original issue**:
It appears to be awesome-specific only.

More precisely it seems to be caused by repeated executions of
`dex-autostart -a -e XFCE`
(https://github.com/QubesOS/qubes-desktop-linux-awesome/blob/master/awesome-3.5.5-dex-autostart.patch)
during awesome restarts.

Btw my dispVMs started working again after the apps were started again.

**Comment by kalkin** [comment]
@3hhh > More precisely it seems to be caused by repeated executions of

Ohh you mean that error. :) I know about it and this is on my agenda, but it's low priority, because it's not part of the main Qubes project.

> I'm a tester, not a believer...

I think I misunderstood you. I thought you mean the X clients of the dispvm survive reboot. Thanks for opening a new issue.

**Comment by kalkin** [comment]
@3hhh If you have some time, have a look at the [Awesome exit Signal](https://awesomewm.org/apidoc/libraries/awesome.html#exit). It has a parameter for restart. So you can only exit `dex` on startup.

**Comment by kalkin** [comment]
@3hhh Looking more in to the `exit` Signals it seems not to help much. While you find out if awesome is currently restarted on exit, but after restart it will reparse the `rc.lua` and you are back where you started ☹

**Comment by cyrinux** [comment]
Hi, there is the same systray problem with i3. And gui not recevered after a X crash due to low memory for example and relogin.

**Comment by 3hhh** [comment]
This fixes it and was tested to work at least with 4.0rc1 even after system crashes (power removed):

```lua
--returns true, if the awesome autostart should run (didn't run before already)
local function autostartRequired()
	local lockFilePath = "/tmp/awesome-autostart"
	local lockFile = io.open(lockFilePath,"r")
		if lockFile == nil then

			--create the file
			lockFile = io.open(lockFilePath,"w")
			io.close(lockFile)

			return true
		else
			io.close(lockFile)
			return false
		end
end

--run autostart programs on first boot (awesome restarts won't trigger this)
if autostartRequired() then
	-- Use dex to run the xdg autostart files
	-- for the moment run the ones for XFCE
	awful.util.spawn_with_shell("dex-autostart -a -e XFCE")

	--custom autostart
	-- [...]
end
```

**Comment by 3hhh** [comment]
The above just didn't work with an X server restart. :-(

Maybe some pgrep solution is more reliable...

**Comment by 3hhh** [comment]
The following appears to work:

```lua
-- Use dex to run the xdg autostart files
-- for the moment run the ones for XFCE
-- NOTE: awful.util.spawn_with_shell() does some really unpredictable stuff wrt quoting --> you are better off with calling spawn() with your favorite shell and controlling the quoting yourself; still it may not work in awesome, even if it works in your shell
awful.util.spawn('bash -c "pgrep -f -a mqui.tray | grep -v $$ || dex-autostart -a -e XFCE"')
```

Even minor changes may break it again though due to the unreliable nature of both spawn() and spawn_with_shell() (the latter appears to be even worse).

**Comment by kalkin** [comment]
@3hhh Thanks. I will integrate this in my Awesome 4.1 patches QubesOS/qubes-desktop-linux-awesome#10

**Comment by 3hhh** [comment]
For ref: Currently I'm using the method depicted in #3626 to do the autostart.

**Comment by 3hhh** [comment]
My workaround did its job for a year or so now.

Since one or two dom0 updates back however every time the X server is restarted I get a popup saying:
XFCE PolicyKit Agent
GDBus.Error:org.freedesktop.PolicyKit1.Error.Failed: An authentication agent already exists for the given subject

It doesn't appear to have any relevance, but is annoying.

Overall my X server restarts appear to be caused by my docking station (driver issue?). It is quite interesting to see that something as simple as docking a laptop can cause the X server to die/restart and thus potentially leave the screen unprotected (xscreensaver dies with it, but doesn't restart).

**Comment by kalkin** [comment]
```
XFCE PolicyKit Agent
GDBus.Error:org.freedesktop.PolicyKit1.Error.Failed: An authentication agent already exists for the given subject
```
Not a QubesOS issue. Have the same on Gentoo while using Awesomewm

**Comment by 3hhh** [comment]
> ```
> XFCE PolicyKit Agent
> GDBus.Error:org.freedesktop.PolicyKit1.Error.Failed: An authentication agent already exists for the given subject
> ```
> 
> Not a QubesOS issue. Have the same on Gentoo while using Awesomewm

No, it was caused by my workaround, which didn't work correctly anymore as the Qubes applet binary names changed.

So currently I'm using
`
pgrep -fa qui-devices | grep -v $$ || ( dex-autostart -a -e XFCE < /dev/null &> /dev/null ; sleep 10 )
`

Anyway this confirms that the originally reported bug still exists in Qubes 4.1.
